### DISCUSSION TITLE
How does Oceananigans deal with topography?

### DISCUSSION THREAD

**Discussion by fpersan** [CliMA/Oceananigans.jl#2410] [Experimental features]
Hi, I am very interested in learning Oceananigans, but I am working on problems dealing with topography on a z-coordinate model (MItgcm). MITgcm uses partial-step cells to work around pressure gradient errors. It came to my attention that the partial-step cell method has been implemented for Oceananigans recently. Has the method been validated already? I am particularly curious to try and run a simple experiment to generate internal lee waves over a rough bottom.

**Comment by francispoulin** [comment]
There is a PR, #2306 , which introduces a partial cell method but there still needs to be some testing before it's merged into main.  Maybe in a week or so?

**Reply by glwagner** [reply]
Note, we also have a "full cell method" for bathymetry now!

Can I move this discussion to "experimental features"? We can add a few examples of how to use the current features for bathymetry, and perhaps outline a vision for future developments. As @francispoulin mentioned we are working on partial cells, but I also think there is scope for shaved cells and such at some point.

**Reply by iuryt** [reply]
Shaved cells would be awesome!

**Reply by fpersan** [reply]
That's great!! I moved it to "experimental features." Shaved cells would be a considerable advancement indeed.

**Reply by tomchor** [reply]
Something to note is that at this point the boundary conditions capabilities associated with these IBM implementations are limited. @whitleyv can correct me if I'm wrong but the full cell IBM that's implemented I think only has no-flux BC for tracers and no-slip BC for momentum. We're experimenting with a drag BC for that implementation in https://github.com/CliMA/Oceananigans.jl/pull/2275 but for now that's still hard-coded.

@francispoulin I'm assuming that status of BCs for your partial cell implementation is similar?

**Comment by glwagner** [comment]
Here's a few places to find examples of working with the current `ImmersedBoundaryGrid` implementation:

* Validation cases (sometimes these go stale, please submit PRs to update them if that's the case): https://github.com/CliMA/Oceananigans.jl/tree/main/validation/immersed_boundaries
* "Near global quarter degree" script: https://github.com/CliMA/Oceananigans.jl/blob/main/validation/near_global_lat_lon/near-global-quarter-degree.jl
* A forcing function implementation of bottom drag: https://github.com/CliMA/Oceananigans.jl/blob/8e8a9d540af5738dd5eaf3d7180c455315785b61/validation/near_global_lat_lon/near-global-quarter-degree.jl#L169-L177

Finally, there's #2407 to discuss the API for boundary condition fluxes and #2275 for a hard-coded implementation of a drag condition appropriate for LES. There's also some discussion on #2275 on what a more general implementation might look like.

Please don't hesitate to post additional comments if this information goes out of date! And more examples on the internet are always welcome if anyone has links.

**Comment by caim-d** [comment]
Hi, I'm a little new to using Github (and Julia) so sorry if this is made clear somewhere I haven't seen! But just to follow up, if I want to use partial cells rather than full cells, is the way to do this to replace what is written in the documentation, i.e. giving a profile for the hill and then using GridFittedBottom(hill), with PartialCellBottom(hill), or would I need to do something else in addition? I've seen what's written in https://github.com/CliMA/Oceananigans.jl/blob/main/validation/immersed_boundaries/partial_cell_grid_analysis.jl but just want to be sure.

**Reply by tomchor** [reply]
I think technically you just need to substitute `GridFittedBottom` with `PartialCellBottom`, but I'm not sure if the partial cell bottom is validated yet. If I recall correctly @francispoulin found some weird behavior at some point and probably has more information about this. That said, I've used partial cells in the part and didn't really notice anything.

@caim-d if you'd like to spend some effort to validate it, you can create a PR and we'd be glad to help you out with that.

**Reply by caim-d** [reply]
Thank you (and sorry for the slow reply)! What needs to be done to validate the code?

**Reply by tomchor** [reply]
Again I'm not the best person to answer here because I didn't work directly on the implementation of any immersed boundary method. My impression is that the best way to start would be to extend some of the IB examples in the validation (https://github.com/CliMA/Oceananigans.jl/tree/main/validation/immersed_boundaries) to also use `PartialCellBottom`. Then you could see if something weird happens.

I also think it would be good to compare more rigorously with previous results such as [Adcroft et al.](https://journals.ametsoc.org/view/journals/mwre/125/9/1520-0493_1997_125_2293_rotbsc_2.0.co_2.xml) and [something like this](https://link.springer.com/article/10.1007/s10494-018-0005-5).

Finally, we'd need to include a few tests for `PartialCellBottom` since I don't think there are any.

@glwagner @francispoulin @whitleyv @simone-silvestri  you guys worked directly on the implementation on the IBMs we have. Please feel free to chime in.

**Reply by glwagner** [reply]
I would start by setting up the simulation you'd like to use for science and comparing the results between `GridFittedBottom` and `PartialCellBottom` immersed boundaries. You should evaluate the quality of the solution qualitatively (eg by visualizing it), and also quantitatively. One important metric is the tracer budget. You will find there is some leakage across the immersed boundary, but hopefully it is small. It would be interesting to evaluate the differences in leakage between the two immersed boundary methods.

[Adcroft et al 1997](https://journals.ametsoc.org/view/journals/mwre/125/9/1520-0493_1997_125_2293_rotbsc_2.0.co_2.xml) (linked by @tomchor) is a good paper to read to understand the basics of the method. It may also give you some ideas for interesting metrics to evaluate. I wouldn't spend too much time reproducing those results specifically (we have done that). Instead, I would look at your case specifically.

**Reply by caim-d** [reply]
Thank you!
